libs/message-coder/src/types.ts

```diff
@@ -0,0 +1,153 @@
+import { isResult, ok, Result } from '@votingworks/types';
+import { Buffer } from 'buffer';
+
+/**
+ * Errors that can occur when encoding or decoding.
+ */
+export enum CoderError {
+  SmallBuffer = 'SmallBuffer',
+  InvalidValue = 'InvalidValue',
+  UnsupportedOffset = 'UnsupportedOffset',
+}
```

Comment:
Nit, but I've really come to prefer string unions over enums. Easier to type and doesn't require imports. I don't care what it is for this PR or this specific case, but if there are reasons you explicitly prefer `enum` here, what is it? I think we need to outline some best practices internally.

Reply:
I think they're fine for string enums. For numeric enums I don't think they work so well. For example ([playground](https://www.typescriptlang.org/play?target=7#code/PQ18ZXAUCAEBXAzgSwHYHN4AMCm6iAtjvAGYD2ATvIUXlagMbwBuAhgDaJ7LwhxoQ4RBgwCxeACVeABwrpkeAKJUq1AMIUAJnngBvGPH4AqE0eMn4AdXVZy1IuwAu8CmXjOAFnvaKA7gzk6kSePvDITH7oDAB08ABC7NrwfsiBVLEW8CawxgBiAPJSALIAggAqAPrKUlLF8AC88AAMADRixsBm2VYAylHoMTRpGXgpAEQAkugcnKgpTBRETujaE1nGOXnwUwByAGplADJTACJVGoUl5XtnTa0AHgAc7Z2m5lv9g8OpAQzjeDTWZcBbwABSFAARrszhtejt9kdThdwYUElVzg8Wi8AIwdAC+YjIiHQTGcqAU8FkjHQzlU6ioAAolroAFzSOQKJQMzQ6PAASg5rAoYMMxiWigonDwsU4FEwTIA5LV6lI2Uq2pzkPJFCo1HzdABtVl4AC6AoA3DAiYJ4D4qHopvBBkglNrdTyDVQtLpPBR4Jg8K5vHp5shXO5qRRkGgoTK2FweMgYDSMPTvUyZDrufrGb7ZUVSpUanVilaxGIRNXq3bVoguLgJCQHDQ6AxmInuLx+LAa-3oGJJRGPbneT7+QAmB7ij69Gx2bCUKhOSMeUN-dJBMghMJ6SLROKJZKbjKbSw7Ivlaqq4oct4WbqfSzwAaHkb-R2TGZzMFLFZ+Os57bBYSInOclzXLcZz3i8D5dD0Xyvj8QSjAC34gvMKSQjC5zwl8iKHOBqLopiMFPM8+I2qkfDDs41owM4ACesh6Nmnp5oaeDTs0TLMaxUbsWO3oFpOApGgA1ngTFRvxeCCVyerjqJZoMSSZIUlSaZ0uOk4svyHJCUpIlTkKbCiiks4yq42ioDqnDsExHIRrS2AAD5IGseBkBg4wMcY6SoM4TBePA+m6AKBjZC67DukZXr5lOsRXiWt7qtFxi2fZjkPEqKU3mWUhKtaWxbFCjrsBJ-lbFEcWKQlXGTrEYEopBNxlHcbIZfAWWyA5TG5S1EFXO1dzFd15V4JV1USrFbH1ZxE66E1Q0kRi5xdaVmV2X1OXNEqq1VGi61nONW3wJN00WLaErctKsryoqKqFRqWq9f1Fa2ggzoRqgnCcPa7CsHozgBig805sZiXLf6gbBnu8Dhmu0axqg8Z6HMyYdAgyABs6gxKq4QMWX88CoEQ8hUK4qAeM6-h+ETiCg-+fXBngqa0hmjJ6fFi2iclxTXqWaoVkAA)):

```ts
/////////////////////////////////////
// using `enum` for numeric values //
/////////////////////////////////////

enum ResponseErrorCode {
  /**
   * Wrong format of the answer from the scanner. Bad answer.
   */
  FORMAT_ERROR = 0,

  /**
   * Scanner answered "Invalid command".
   */
  INVALID_COMMAND = 0x80,

  /**
   * Scanner answered "Invalid Job ID".
   */
  INVALID_JOB_ID = 0x81,
}

function printError(code: ResponseErrorCode): void {
  console.log('ERROR:', ResponseErrorCode[code]);
}

// here I can use ResponseErrorCode to get the list of possible values
printError(ResponseErrorCode.FORMAT_ERROR);



//////////////////////////////////////
// manual `enum` for numeric values //
//////////////////////////////////////

const ResponseErrorCode2 = {
  /**
   * Wrong format of the answer from the scanner. Bad answer.
   */
  FORMAT_ERROR: 0,

  /**
   * Scanner answered "Invalid command".
   */
  INVALID_COMMAND: 0x80,

  /**
   * Scanner answered "Invalid Job ID".
   */
  INVALID_JOB_ID: 0x81,
} as const;

type ResponseErrorCode2 = (typeof ResponseErrorCode2)[keyof typeof ResponseErrorCode2];

function printError2(code: ResponseErrorCode2): void {
  let display: string | undefined;

  switch (code) {
    case ResponseErrorCode2.FORMAT_ERROR:
      display = 'FORMAT_ERROR';
      break;

    case ResponseErrorCode2.INVALID_COMMAND:
      display = 'INVALID_COMMAND';
      break;

    case ResponseErrorCode2.INVALID_JOB_ID:
      display = 'INVALID_JOB_ID';
      break;
  }

  console.log('ERROR:', display);
}

// I still have to use ResponseErrorCode2 to get the list of possible values,
// so I can't avoid an import if I want autocomplete
printError2(ResponseErrorCode2.FORMAT_ERROR);
```

The manual version retains most of the benefits of the built-in `enum`, but lacks the reverse mapping from value to key.

Reply:
When numeric values are involved, I think an enum makes most sense.

In this particular case, I am thinking that instead of:
```
export enum CoderError {
  SmallBuffer = 'SmallBuffer',
  InvalidValue = 'InvalidValue',
  UnsupportedOffset = 'UnsupportedOffset',
}
```
You could have:
```
export type CoderError = 'SmallBuffer' | 'InvalidValue' | 'UnsupportedOffset'
```
At one point the typescript linting and autocomplete didn't support this fully, but now it does.

Reply:
Switched to the string union style for this particular enum.